tests/server-integration/Common/BootServer.hs

```diff
@@ -0,0 +1,101 @@
+import Control.Exception (SomeException(SomeException))
+import System.Environment (setEnv, unsetEnv, withArgs)
+import System.Time.Extra (sleep)
+import Test.Tasty.HUnit (assertFailure, (@=?))
+import Web.Main (RunServerException(..), runServer)
+
+testPort :: String
+testPort = "8079"
+
+testServer :: String -> [String] -> IO () -> IO () -> (String -> IO ()) -> IO ()
+testServer envPort args serverRunning serverStopped serverCrashed = do
+  setEnv "COFFER_SERVER_PORT" envPort
+  server <- async $ withArgs args runServer
+  sleep 1
+  serverStatus <- poll server
+  case serverStatus of
+    Nothing -> cancel server >> serverRunning
+    Just (Right _) -> serverStopped
+    Just (Left (SomeException err)) -> serverCrashed $ show err
+  unsetEnv "COFFER_SERVER_PORT"
+
+unit_run_with_cmd_option_port :: IO ()
+unit_run_with_cmd_option_port = testServer
+  ""
+  ["--port=" ++ testPort]
+  (pure ())
+  (assertFailure "Server ended it's work")
+  (\err -> assertFailure ("Server raised exception : " ++ show err))
+
+-- can we hide stdout?
+unit_run_with_bad_str_cmd_option_port :: IO()
+unit_run_with_bad_str_cmd_option_port = testServer
+  ""
+  ["--port=abs"]
+  (assertFailure "Server is runnung with bad port env var")
+  (assertFailure "Server successfully ended it's work with bad port env var")
+  (\err -> "ExitFailure 1" @=? err)
+
+unit_run_with_bad_cmd_option_port :: IO()
+unit_run_with_bad_cmd_option_port = testServer
+  ""
+  ["--port=-1"]
+  (assertFailure "Server is runnung with bad port env var")
+  (assertFailure "Server successfully ended it's work with bad port env var")
+  (\err -> show (RunServerIncorrectPort -1) @=? err)
+
+unit_run_with_env_var_port :: IO()
+unit_run_with_env_var_port = testServer
+  testPort
+  []
+  (pure ())
+  (assertFailure "Server ended it's work")
+  (\err -> assertFailure $ "Server raised exception : " ++ show err)
+
+unit_run_with_bad_env_var_port :: IO()
+unit_run_with_bad_env_var_port = testServer
+  "bad_port"
+  []
+  (assertFailure "Server is runnung with bad port env var")
+  (assertFailure "Server successfully ended it's work with bad port env var")
+  (\err -> show (RunServerEnvVarParseFail "Prelude.read: no parse") @=? err)
+
+unit_run_with_no_port_specified :: IO()
+unit_run_with_no_port_specified =  testServer
+  ""
+  []
+  (assertFailure "Server is runnung without port specified")
+  (assertFailure "Server successfully ended it's work without port specified")
+  (\err -> show RunServerNoPortSpecified @=? err)
+
+unit_run_with_cmd_opt_and_bad_env_var_port :: IO()
+unit_run_with_cmd_opt_and_bad_env_var_port = testServer
+  "-1"
+  ["--port=" ++ testPort]
+  (pure ())
+  (assertFailure "Server ended it's work")
+  (\err -> assertFailure $ "Server raised exception : " ++ show err)
+
+
+unit_run_with_bad_cmd_opt_and_env_var_port :: IO()
+unit_run_with_bad_cmd_opt_and_env_var_port = testServer
+  testPort
+  ["--port=-1"]
+  (assertFailure "Server made prefernce to env var instead of cmd option")
```

Comment:
"prefernce" -> "preference"

But a better way of writing error messages in tests is to state what you expected to happen and what actually happened. E.g.

> Expected command-line option to take precedence over the env var, but the env var was used instead.

This way, if this tests ever starts failing in the future, it'll be much easier for the person troubleshooting the issue to figure out how _it's supposed to work_ and how it differs from the actual behavior.

Reply:
done